### PR TITLE
Groundwork for venue feature

### PR DESCRIPTION
**Pull request by nseah21**
### Changes made:
1. `Booking` classes and logic were added. Note the difference between an `AdHocBooking` and a `RecurrentBooking`.
2. Unit tests were added for each booking subclass.
3. A `VenueTableView` was added to allow the user to view venue data in a separate table. 
4. The UI for MainWindow was modified to use a `TabPane` with two `Tabs`, each containing the `ResidentTableView` and `VenueTableView` respectively.
5. `LogicManager` and `ModelManager` were modified to work with a venue list, so that `VenueTableView` could be hooked up to `MainWindow`
6. A simple `Venue` class was also added. This needs work.

### To-do:
1. Add fields in Venue for displaying in the table.  
2. Link `Venue` and the venue list to the storage.
3. Add commands to allow booking of venues. 
    - One possible format is `book /a venue_id resident_matric_number date start_time end_time` for an `AdHocBooking`, and
    - `book /r venue_id resident_matric_number day_of_week start_time end_time` for a `RecurrentBooking`
    - We can explore other ways to identify a resident when booking a venue 😄 
4. Update documentation, user guide and testing for the above three bullet points. 

### Some comments:
1. The switching between tabs in the UI can be achieved in multiple ways: 
    - Using the left-side of the keyboard `Ctrl+Tab` 
    - Using the right-side of the keyboard `Ctrl+PgUp` or `Ctrl+PgDn`
    - Using the mouse to click each tab
    - We can discuss if we need to implement a command to switch tabs 😃

### CODE REVIEW

File: src/main/java/seedu/rc4hdb/model/venues/AdHocBooking.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+                    && this.startTime.isBefore(otherBooking.endTime)
+                    || this.endTime.isAfter(otherBooking.startTime)
+                    && this.endTime.isBefore(otherBooking.endTime);
+            boolean isOverlappingOrSuperset = otherBooking.startTime.isAfter(this.startTime)
+                    && otherBooking.startTime.isBefore(this.endTime)
+                    || otherBooking.endTime.isAfter(this.startTime)
+                    && otherBooking.endTime.isBefore(this.endTime);
+            return otherBooking.date.equals(this.date)
+                    && (isOverlappingOrSubset || isOverlappingOrSuperset);
+        }
```
Comment (alvinjiang1) [inline src/main/java/seedu/rc4hdb/model/venues/AdHocBooking.java:60]: This method looks quite important in the future, so we could potentially protect it by adding some assert statements?

File: src/main/java/seedu/rc4hdb/model/venues/RecurrentBooking.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        }
+        boolean isOverlappingOrSubset = this.startTime.isAfter(booking.startTime)
+                && this.startTime.isBefore(booking.endTime)
+                || this.endTime.isAfter(booking.startTime)
+                && this.endTime.isBefore(booking.endTime);
+        boolean isOverlappingOrSuperset = booking.startTime.isAfter(this.startTime)
+                && booking.startTime.isBefore(this.endTime)
+                || booking.endTime.isAfter(this.startTime)
+                && booking.endTime.isBefore(this.endTime);
+        return isOverlappingOrSubset || isOverlappingOrSuperset;
```
Comment (alvinjiang1) [inline src/main/java/seedu/rc4hdb/model/venues/RecurrentBooking.java:51]: Same as above for this comment, some defensive programming tactics could be useful here :)

**Review by alvinjiang1 — APPROVED** [review]
LGTM 👌 Great job, UI seems to work!

File: src/main/java/seedu/rc4hdb/model/venues/Booking.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+    protected int id;
+
+    /**
+     * Constructor for a Booking instance.
+     * @param resident The resident trying to book the venue.
+     * @param startTime The start time of the booking.
+     * @param endTime The end time of the booking.
+     */
+    public Booking(Resident resident, String startTime, String endTime) {
```
Comment (nealetham) [inline src/main/java/seedu/rc4hdb/model/venues/Booking.java:27]: Is it possible that users will input a `startTime`, and `endTime` that is nonsensical? If so, we could make an exception class for all venue booking purposes, and use it here.

File: src/test/java/seedu/rc4hdb/model/venues/AdHocBookingTest.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    @Test
+    public void adHocBooking_doesNotClashWith_mutuallyExclusiveAdHocBooking() {
+        assertFalse(SEPT_20_2022_5_TO_10_PM.clashesWith(SEPT_20_2099_7_TO_11_PM));
+    }
+
+    @Test
+    public void adHocBooking_clashesWith_overlappingRecurrentBooking() {
+        assertTrue(TUESDAY_730_TO_9PM.clashesWith(SEPT_20_2022_5_TO_10_PM));
+    }
+}
```
Comment (nealetham) [inline src/test/java/seedu/rc4hdb/model/venues/AdHocBookingTest.java:44]: Once we create the exception classes as mentioned above, we could add some negative testing here, and in the `RecurrentBookingTest` class

**Review by nealetham — APPROVED** [review]
Minor nitpicks but good overall! 👍

File: src/main/java/seedu/rc4hdb/model/venues/Booking.java (hunk 3 above)
Comment (nseah21) [reply src/main/java/seedu/rc4hdb/model/venues/Booking.java:27]: Thanks for pointing that out Neale. The original intention was to enforce these checks in a `Parser` class after `BookCommand` is implemented. That said, I certainly agree that having a `VenueException` or `BookingException` class will be useful. 😄

File: src/main/java/seedu/rc4hdb/model/venues/AdHocBooking.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        return currentDateTime.isAfter(expiryDateTime);
+    }
+
+    /**
+     * Implements the clashesWith() method in the Booking class.
+     * @param booking The booking to check for clashes with.
+     * @return True if both bookings clash.
+     */
+    @Override
+    public boolean clashesWith(Booking booking) {
```
[the remark this answers is not in the document]
Comment (nseah21) [reply src/main/java/seedu/rc4hdb/model/venues/AdHocBooking.java:44]: This is a good suggestion. The `equals` method should be added for both classes.

File: src/main/java/seedu/rc4hdb/model/venues/RecurrentBooking.java (hunk 2 above)
Comment (nseah21) [reply src/main/java/seedu/rc4hdb/model/venues/RecurrentBooking.java:51]: Good idea Alvin, will work on that!

**Pull request merged**